File: backend/src/services/auth-service.ts
import type { PoolClient } from "pg";
import { config } from "../config";
import { query } from "../db";
import { ConflictError, UnauthorizedError } from "../errors";
import {
  SQL_CHECK_EMAIL_EXISTS,
  SQL_DELETE_SESSION_BY_ID,
  SQL_DELETE_SESSION_BY_TOKEN,
  SQL_INSERT_SESSION,
  SQL_INSERT_USER,
  SQL_SELECT_SESSION_BY_TOKEN,
  SQL_SELECT_USER_FOR_LOGIN,
  SQL_SELECT_USER_IS_ACTIVE,
} from "../sql/auth";
import { SQL_CHECK_ORG_SLUG_EXISTS, SQL_INSERT_MEMBERSHIP, SQL_INSERT_ORG } from "../sql/orgs";
import {
  createAccessToken,
  createRefreshToken,
  hashPassword,
  verifyPassword,
} from "../utils/security";

function slugify(value: string): string {
  const slug = value.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return slug.slice(0, 50) || "org";
}

async function uniqueSlug(client: PoolClient, base: string): Promise<string> {
  let slug = base;
  let counter = 1;
  while (true) {
    const result = await client.query<{ id: string }>(SQL_CHECK_ORG_SLUG_EXISTS, [slug]);
    if (!result.rowCount) return slug;
    slug = `${base}-${counter}`;
    counter += 1;
  }
}

async function buildTokenResponse(userId: string) {
  return {
    access_token: await createAccessToken({ sub: userId }),
    token_type: "bearer",
    expires_in: config.accessTokenExpireMinutes * 60,
  };
}

export async function registerUser(
  client: PoolClient,
  payload: { email: string; password: string; full_name?: string | null; org_name?: string | null },
  meta: { userAgent?: string; ipAddress?: string }
) {
  const existing = await client.query<{ id: string }>(SQL_CHECK_EMAIL_EXISTS, [payload.email]);
  if (existing.rowCount) throw new ConflictError("Email already registered");

  const hashedPassword = await hashPassword(payload.password);
  const userResult = await client.query<{ id: string }>(SQL_INSERT_USER, [
    payload.email, hashedPassword, payload.full_name ?? null,
  ]);
  const userId = userResult.rows[0].id;

  const orgName = payload.org_name ?? `${payload.full_name ?? payload.email.split("@")[0]}'s Workspace`;
  const orgSlug = await uniqueSlug(client, slugify(orgName));
  const orgResult = await client.query<{ id: string }>(SQL_INSERT_ORG, [orgSlug, orgName]);
  const orgId = orgResult.rows[0].id;

  await client.query(SQL_INSERT_MEMBERSHIP, [orgId, userId, "owner"]);

  const refreshToken = createRefreshToken();
  await client.query(SQL_INSERT_SESSION, [
    userId, refreshToken, meta.userAgent ?? null, meta.ipAddress ?? null,
    String(config.refreshTokenExpireDays),
  ]);

  return { tokenResponse: await buildTokenResponse(userId), refreshToken };
}

export async function loginUser(
  client: PoolClient,
  payload: { email: string; password: string },
  meta: { userAgent?: string; ipAddress?: string }
) {
  const result = await client.query<{ id: string; hashed_password: string | null; is_active: boolean }>(
    SQL_SELECT_USER_FOR_LOGIN, [payload.email]
  );
  const user = result.rows[0];
  if (!user?.hashed_password) throw new UnauthorizedError("Invalid credentials");
  if (!(await verifyPassword(payload.password, user.hashed_password))) throw new UnauthorizedError("Invalid credentials");
  if (!user.is_active) throw new UnauthorizedError("Account is disabled");

  const refreshToken = createRefreshToken();
  await client.query(SQL_INSERT_SESSION, [
    user.id, refreshToken, meta.userAgent ?? null, meta.ipAddress ?? null,
    String(config.refreshTokenExpireDays),
  ]);

  return { tokenResponse: await buildTokenResponse(user.id), refreshToken };
}

export async function refreshSession(
  client: PoolClient,
  oldRefreshToken: string,
  meta: { userAgent?: string; ipAddress?: string }
) {
  const sessionResult = await client.query<{ id: string; user_id: string; expires_at: string }>(
    SQL_SELECT_SESSION_BY_TOKEN, [oldRefreshToken]
  );
  const session = sessionResult.rows[0];
  if (!session || new Date(session.expires_at) < new Date()) {
    throw new UnauthorizedError("Invalid or expired refresh token");
  }

  const userResult = await client.query<{ id: string; is_active: boolean }>(
    SQL_SELECT_USER_IS_ACTIVE, [session.user_id]
  );
  const user = userResult.rows[0];
  if (!user || !user.is_active) throw new UnauthorizedError("User not found or inactive");

  await client.query(SQL_DELETE_SESSION_BY_ID, [session.id]);

  const refreshToken = createRefreshToken();
  await client.query(SQL_INSERT_SESSION, [
    user.id, refreshToken, meta.userAgent ?? null, meta.ipAddress ?? null,
    String(config.refreshTokenExpireDays),
  ]);

  return { tokenResponse: await buildTokenResponse(user.id), refreshToken };
}

export async function logoutSession(refreshToken: string) {
  await query(SQL_DELETE_SESSION_BY_TOKEN, [refreshToken]);
}

File: backend/app/services/auth_service.py
from __future__ import annotations

import re
from typing import Any

from app.core.config import settings
from app.errors import ConflictError, UnauthorizedError
from app.sql import auth as sql_auth
from app.sql import orgs as sql_orgs
from app.utils.security import create_access_token, create_refresh_token, hash_password, verify_password


def _slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    return slug[:50] or "org"


async def _unique_slug(conn, base: str) -> str:
    slug = base
    counter = 1
    while True:
        row = await conn.fetchrow(sql_orgs.check_org_slug_exists(slug))
        if row is None:
            return slug
        slug = f"{base}-{counter}"
        counter += 1


def _build_token_response(user_id: str) -> dict[str, Any]:
    return {
        "access_token": create_access_token({"sub": user_id}),
        "token_type": "bearer",
        "expires_in": settings.access_token_expire_minutes * 60,
    }


async def register_user(
    conn,
    payload: dict[str, Any],
    meta: dict[str, Any],
) -> dict[str, Any]:
    existing = await conn.fetchrow(sql_auth.check_email_exists(payload["email"]))
    if existing:
        raise ConflictError("Email already registered")

    hashed = hash_password(payload["password"])
    user_row = await conn.fetchrow(
        sql_auth.insert_user(payload["email"], hashed, payload.get("full_name"))
    )
    user_id = str(user_row["id"])

    org_name = (
        payload.get("org_name")
        or f"{payload.get('full_name') or payload['email'].split('@')[0]}'s Organization"
    )
    org_slug = await _unique_slug(conn, _slugify(org_name))
    org_row = await conn.fetchrow(sql_orgs.insert_org(org_slug, org_name))
    org_id = str(org_row["id"])

    await conn.execute(sql_orgs.insert_membership(org_id, user_id, "owner"))

    refresh_token = create_refresh_token()
    await conn.execute(
        sql_auth.insert_session(
            user_id=user_id,
            refresh_token=refresh_token,
            user_agent=meta.get("user_agent"),
            ip_address=meta.get("ip_address"),
            expires_days=settings.refresh_token_expire_days,
        )
    )

    return {"token_response": _build_token_response(user_id), "refresh_token": refresh_token}


async def login_user(
    conn,
    payload: dict[str, Any],
    meta: dict[str, Any],
) -> dict[str, Any]:
    row = await conn.fetchrow(sql_auth.select_user_for_login(payload["email"]))
    if not row or not row["hashed_password"]:
        raise UnauthorizedError("Invalid credentials")
    if not verify_password(payload["password"], row["hashed_password"]):
        raise UnauthorizedError("Invalid credentials")
    if not row["is_active"]:
        raise UnauthorizedError("Account is disabled")

    user_id = str(row["id"])
    refresh_token = create_refresh_token()
    await conn.execute(
        sql_auth.insert_session(
            user_id=user_id,
            refresh_token=refresh_token,
            user_agent=meta.get("user_agent"),
            ip_address=meta.get("ip_address"),
            expires_days=settings.refresh_token_expire_days,
        )
    )

    return {"token_response": _build_token_response(user_id), "refresh_token": refresh_token}


async def refresh_session(
    conn,
    old_refresh_token: str,
    meta: dict[str, Any],
) -> dict[str, Any]:
    from datetime import datetime, timezone

    session = await conn.fetchrow(sql_auth.select_session_by_token(old_refresh_token))
    if not session:
        raise UnauthorizedError("Invalid or expired refresh token")

    expires_at = session["expires_at"]
    if hasattr(expires_at, "tzinfo") and expires_at.tzinfo:
        if expires_at < datetime.now(timezone.utc):
            raise UnauthorizedError("Invalid or expired refresh token")

    user = await conn.fetchrow(sql_auth.select_user_is_active(session["user_id"]))
    if not user or not user["is_active"]:
        raise UnauthorizedError("User not found or inactive")

    await conn.execute(sql_auth.delete_session_by_id(session["id"]))

    user_id = str(user["id"])
    refresh_token = create_refresh_token()
    await conn.execute(
        sql_auth.insert_session(
            user_id=user_id,
            refresh_token=refresh_token,
            user_agent=meta.get("user_agent"),
            ip_address=meta.get("ip_address"),
            expires_days=settings.refresh_token_expire_days,
        )
    )

    return {"token_response": _build_token_response(user_id), "refresh_token": refresh_token}


async def logout_session(conn, refresh_token: str) -> None:
    await conn.execute(sql_auth.delete_session_by_token(refresh_token))
